dashboard/src/utils/dateFormatter.js

```diff
@@ -1,3 +1,8 @@
+import { isValidDate } from "@patternfly/react-core"
+
 export const formatDate=(date)=>{
-    return date.toISOString().slice(0,10)
+    if(!isValidDate(date))
+    return;
+    const dateWithoutOffset=date.toString().split(/\+|-/)
```

Comment:
Isn't the logic already centralized?I mean any changes we want to do with how we are processing offset is only in this file.Am I missing something here?

Reply:
Ha; good point. I lost track of the file I was reading. 😆

Reply:
> all timestamps you get from the server ought to be interpreted as UTC

@dbutenhof, do they come with a timezone designator?  (E.g., do they end in `Z` or in something like `+00` or `-00:00` or `+0000`?)

> whereas I think the dashboard ought to be operating in the default browser localized timezone

Javascript pretty much imposes this.

> or possibly, in the future, a specific server/dashboard user preference setting

This might be hard, but, I'm hoping we can obviate the need/desire for it.

> which means we'd want to convert timestamps to the active timezone rather than dropping that information with a string operation.

The problem is that there is context relative to some of the artifacts which is not necessarily related to that of the current user.  For instance, result names currently have a timestamp in them; if we localize the "created at" timestamp, then it won't necessarily match the name of the result when viewed (e.g., in India).  Analogously, if someone in Eastern Time says "look at this timestamp", it will be hard for someone in India to find if both users see things tagged in their respective local times.

On the other hand, if we represent everything in UTC internally and display everything in UTC, those sorts of problems go away.  But, we are left with the fact that Patternfly (etc.) collects times from the user in local time, and we have to catch that and convert them to UTC properly.

The case in point is that Shubham is implementing a range picker which displays all the runs between a starting calendar _date_ and an ending calendar _date_.  Internally, these are converted to date/times with the time of midnight at the start of the indicated day.  All good so far, but if Shubham, in India, selects an end-time of `2022-04-27`, this gets truncated to `2022-04-27T00:00:00+05:30`, which gets represented as `2022-04-26:18:30:00Z` and then any results from 4/27 disappear....

So, we need to establish the understanding that the User is to use the Patternfly date picker to select a UTC-based date, not a local date; and we need to take the result produced by Patternfly and convert or adjust it appropriately.  In the success case, Patternfly produces both a string and a JS `Date` object.  While we could take the `Date` and apply the timezone offset to it to "fix" it, we could also take the string, put it into a UTC  context, and then produce a `Date` from that.

Reply:
> > all timestamps you get from the server ought to be interpreted as UTC
> 
> @dbutenhof, do they come with a timezone designator? (E.g., do they end in `Z` or in something like `+00` or `-00:00` or `+0000`?)

Nope, it's currently entirely implicit, and I need to change that. (Probably sooner rather than later.)

> > whereas I think the dashboard ought to be operating in the default browser localized timezone
> 
> Javascript pretty much imposes this.
> 
> > or possibly, in the future, a specific server/dashboard user preference setting
> 
> This might be hard, but, I'm hoping we can obviate the need/desire for it.

As you point out below, communicating timestamps across time zones can be awkward; being able to use the Pbench dashboard in UTC (or IST) even though your normal browser preference is EST, could be a big advantage.

> > which means we'd want to convert timestamps to the active timezone rather than dropping that information with a string operation.
> 
> The problem is that there is context relative to some of the artifacts which is not necessarily related to that of the current user. For instance, result names currently have a timestamp in them; if we localize the "created at" timestamp, then it won't necessarily match the name of the result when viewed (e.g., in India). Analogously, if someone in Eastern Time says "look at this timestamp", it will be hard for someone in India to find if both users see things tagged in their respective local times.

"Time" is absolute; sure, the original timestamp may be UTC or IST, but showing that timestamp in CDT is perfectly legitimate if the user is in Texas. If you really want to use the same timestamp representation when communicating between two timezones, you can both choose to show them in UTC, and maybe a lot of people will want that to be the default... my point is that it's unreasonable to impose that restriction.

> On the other hand, if we represent everything in UTC internally and display everything in UTC, those sorts of problems go away. But, we are left with the fact that Patternfly (etc.) collects times from the user in local time, and we have to catch that and convert them to UTC properly.

I think that's backwards.

> The case in point is that Shubham is implementing a range picker which displays all the runs between a starting calendar _date_ and an ending calendar _date_. Internally, these are converted to date/times with the time of midnight at the start of the indicated day. All good so far, but if Shubham, in India, selects an end-time of `2022-04-27`, this gets truncated to `2022-04-27T00:00:00+05:30`, which gets represented as `2022-04-26:18:30:00Z` and then any results from 4/27 disappear....

We definitely need to do time zone normalizations *correctly* and *consistently*, but that doesn't mean we shouldn't do them.

Reply:
> it's currently entirely implicit, and I need to change that. (Probably sooner rather than later.)

Ugh...then @shubham-html-css-js will need to add it until the server can.

> I think that's backwards.

The problem is, if someone in India wants to look at a dataset which someone in Eastern Time collected on a particular day, then there is a huge challenge if the date picker is working in local time.

Granted, if we had the ability for the person in India to set the UI to Eastern Time, then this would be easy to solve.  However, I don't foresee us having that for quite some time.  In the interim, I think it would make things most usable if we declared that all displays and inputs are to be interpreted as UTC.

Reply:
> > it's currently entirely implicit, and I need to change that. (Probably sooner rather than later.)
> 
> Ugh...then @shubham-html-css-js will need to add it until the server can.
> 
> > I think that's backwards.
> 
> The problem is, if someone in India wants to look at a dataset which someone in Eastern Time collected on a particular day, then there is a huge challenge if the date picker is working in local time.

Sure, and if we'd all just set our clocks and calendars to UTC, scheduling meetings would be a lot easier. However, people (even performance engineers, presumably) also need to interact with the real world, so there are advantages in being able to see things in local time. It's kinda how we do things naturally. While not without issues, the fact that we still have time zones seems to provide an argument for making it work. 😀 

With that said, I don't think we need to get the whole universe (much less the time-space continuum) straightened out for this PR, and in the short term doing everything in UTC may be easiest.

Reply:
I have converted all dates in UTC in my latest commits

Reply:
Can we move this to a follow-up PR?

Reply:
Not really.  See [my previous comment](https://github.com/distributed-system-analysis/pbench/pull/2733#discussion_r862843082).